File: leetcode/java/20_ValidParenthesis.java
class Solution {
    public boolean isValid(String s) {
        if (s.length() % 2 != 0) {
            return false;
        }

        Deque<Character> stack = new ArrayDeque<>();
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (ch == '(' || ch == '{' || ch == '[') {
                stack.addFirst(ch);
                continue;
            }

            if (stack.isEmpty()) {
                return false;
            }

            char prev = stack.removeFirst();
            if (ch == ')' && prev != '(') {
                return false;
            }
            if (ch == '}' && prev != '{') {
                return false;
            }
            if (ch == ']' && prev != '[') {
                return false;
            }
        }

        return stack.isEmpty();
    }
}

File: leetcode/go/20_valid_parenthesis.go
package main

func isValid(s string) bool {
	stack := make([]rune, 0)

	for _, c := range s {
		if c == '(' || c == '{' || c == '[' {
			stack = append(stack, c)
		} else {
			if len(stack) == 0 {
				return false
			}

			switch c {
			case ')':
				if stack[len(stack)-1] == '(' {
					stack = stack[:len(stack)-1]
				} else {
					return false
				}
			case '}':
				if stack[len(stack)-1] == '{' {
					stack = stack[:len(stack)-1]
				} else {
					return false
				}
			case ']':
				if stack[len(stack)-1] == '[' {
					stack = stack[:len(stack)-1]
				} else {
					return false
				}
			}
		}
	}

	return len(stack) == 0
}
